test/samplers/delayed_sampling.js

```diff
@@ -0,0 +1,223 @@
+// in compliance with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+
+import { assert } from 'chai';
+import * as opentracing from 'opentracing';
+import { SAMPLER_API_V2 } from '../../src/samplers/constants';
+import Span from '../../src/span';
+import Utils from '../../src/util';
+import ConstSampler from '../../src/samplers/const_sampler';
+import { adaptSamplerOrThrow } from '../../src/samplers/_adapt_sampler';
+import InMemoryReporter from '../../src/reporters/in_memory_reporter';
+import Tracer from '../../src/tracer';
+
+describe('delayed sampling', () => {
+  declare type Matcher = {
+    tagValue: string,
+    // can add more flags here, like firehose
+  };
+
+  class TagEqualsSampler implements Sampler {
+    apiVersion = SAMPLER_API_V2;
+
+    _tagKey: string;
+    _matchers: { [string]: Matcher };
+    _undecided: SamplingDecision;
+
+    constructor(tagKey: string, matchers: Array<Matcher>) {
+      this._tagKey = tagKey;
+      this._matchers = {};
+      matchers.forEach(m => {
+        this._matchers[m.tagValue] = m;
+      });
+      this._undecided = { sample: false, retryable: true, tags: null };
+    }
+
+    _findTag(tags: Array<Tag>): ?Tag {
+      for (let i = 0; i < tags.length; i++) {
+        if (tags[i].key === this._tagKey) {
+          return tags[i];
+        }
+      }
+      return null;
+    }
+
+    _createOutTags(tagValue: string): { [string]: string } {
+      return {
+        'sampler.type': 'TagEqualsSampler',
+        'sampler.param': tagValue,
+      };
+    }
+
+    _decide(tagValue: any): SamplingDecision {
+      const match: ?Matcher = this._matchers[tagValue];
+      if (match) {
+        return { sample: true, retryable: false, tags: this._createOutTags(match.tagValue) };
+      }
+      return this._undecided;
+    }
+
+    onCreateSpan(span: Span): SamplingDecision {
+      const tag: ?Tag = this._findTag(span.getTags());
+      if (tag) {
+        return this._decide(tag.value);
+      }
+      return this._undecided;
+    }
+
+    onSetOperationName(span: Span, operationName: string): SamplingDecision {
+      return this.onCreateSpan(span);
+    }
+
+    onSetTag(span: Span, key: string, value: any): SamplingDecision {
+      if (key === this._tagKey) {
+        return this._decide(value);
+      }
+      return this._undecided;
+    }
+
+    toString(): string {
+      return 'TagEqualsSampler';
+    }
+
+    close(callback: ?Function): void {
+      if (callback) {
+        callback();
+      }
+    }
+  }
+
+  declare type PrioritySamplerState = {
+    samplerFired: Array<boolean>,
```

Comment:
How is `samplerFired` different than `retryable`?

Reply:
samplerFired is set once the given sampler returns retryabe=false. I will add a type comment.